File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<TreeNode> ans = new ArrayList<>();
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        root = dfs(root,to_delete);
        if (root != null)
            ans.add(root);
        return ans;
    }
    public TreeNode dfs(TreeNode node, int[] to_delete){
        if (node == null)
            return null;
        node.left = dfs(node.left,to_delete);
        node.right = dfs(node.right,to_delete);
        for (int i : to_delete){
            if (node.val == i){
                if (node.left != null)
                    ans.add(node.left);
                if (node.right != null)
                    ans.add(node.right);
                return null;
            }
        }
        return node;
    }

}

File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        ans = []
        def dfs(node):
            if not node:
                return node
            node.left = dfs(node.left)
            node.right  = dfs(node.right)
            if node.val in to_delete:
                if node.left:
                    ans.append(node.left)
                if node.right:
                    ans.append(node.right)
                return None
            return node
        root = dfs(root)
        if root:
            ans.append(root)
        return ans